controllers/hyperparametertuningjob/hyperparametertuningjob_controller_test.go

```diff
@@ -18,1680 +18,538 @@ package hyperparametertuningjob
-			Name:      job.ObjectMeta.Name,
-		}, job)
-		Expect(err).ToNot(HaveOccurred())
+		Context("HasDeletionTimestamp", func() {
 
-		Expect(job.Status.BestTrainingJob.TrainingJobName).ToNot(BeNil())
-		Expect(*job.Status.BestTrainingJob.TrainingJobName).To(Equal(expectedTrainingJobName))
-	})
+			BeforeEach(func() {
+				shouldHaveDeletionTimestamp = true
+				shouldHaveFinalizer = true
+			})
 
-	It("should update the status TrainingJobStatusCounters", func() {
-		// Set status
-		var completed int64 = 1
-		var inProgress int64 = 2
-		var nonRetryableError int64 = 3
-		var retryableError int64 = 4
-		var stopped int64 = 5
-
-		// Set mock error values.
-		description.TrainingJobStatusCounters = &sagemaker.TrainingJobStatusCounters{
-			Completed:         &completed,
-			InProgress:        &inProgress,
-			NonRetryableError: &nonRetryableError,
-			RetryableError:    &retryableError,
-			Stopped:           &stopped,
-		}
+			It("Removes finalizer and deletes HyperParameterTuningJob", func() {
+				ExpectHyperParameterTuningJobToBeDeleted(tuningJob)
+			})
 
-		// Setup mock responses.
-		sageMakerClient := builder.
-			AddDescribeHyperParameterTuningJobResponse(description).
-			Build()
+			It("Requeues after interval", func() {
+				ExpectRequeueAfterInterval(reconcileResult, reconcileError, pollDuration)
+			})
+		})
 
-		// Create controller with mocks and its request
-		controller := createHpoReconciler(k8sClient, CreateMockSageMakerClientProvider(sageMakerClient), createMockHpoTrainingJobSpawnerProvider(noopHpoTrainingJobSpawner{}), CreateMockAwsConfigLoader(), "1s")
-		request := CreateReconciliationRequest(job.ObjectMeta.Name, job.ObjectMeta.Namespace)
+		Context("!HasDeletionTimestamp", func() {
+			BeforeEach(func() {
+				mockSageMakerClientBuilder.
+					AddCreateHyperParameterTuningJobResponse(sagemaker.CreateHyperParameterTuningJobOutput{}).
+					AddDescribeHyperParameterTuningJobResponse(CreateDescribeOutputWithOnlyStatus(sagemaker.HyperParameterTuningJobStatusInProgress))
 
-		// Run test
-		_, err := controller.Reconcile(request)
+				shouldHaveDeletionTimestamp = false
+				shouldHaveFinalizer = true
+			})
 
-		Expect(err).ToNot(HaveOccurred())
+			It("Creates a HyperParameterTuningJob", func() {
 
-		// Get job so we can verify it has the correct status.
-		err = k8sClient.Get(context.Background(), types.NamespacedName{
-			Namespace: job.ObjectMeta.Namespace,
-			Name:      job.ObjectMeta.Name,
-		}, job)
-		Expect(err).ToNot(HaveOccurred())
+				req := receivedRequests.Front().Next().Value
+				Expect(req).To(BeAssignableToTypeOf((*sagemaker.CreateHyperParameterTuningJobInput)(nil)))
 
-		Expect(job.Status.TrainingJobStatusCounters).ToNot(BeNil())
+				createdRequest := req.(*sagemaker.CreateHyperParameterTuningJobInput)
+				Expect(*createdRequest.HyperParameterTuningJobName).To(Equal(controllers.GetGeneratedJobName(tuningJob.ObjectMeta.GetUID(), tuningJob.ObjectMeta.GetName(), MaxHyperParameterTuningJobNameLength)))
```

Comment:
This code is duplicated, why not pull out into function (like `ExpectRequestToStopHyperParameterTuningJob`)?

Reply:
This will improve readability as well.

Reply:
Pull what out into which function?

Reply:
I refer to the following lines:

```go
req := receivedRequests.Front().Next().Value
Expect(req).To(BeAssignableToTypeOf((*sagemaker.CreateHyperParameterTuningJobInput)(nil)))

createdRequest := req.(*sagemaker.CreateHyperParameterTuningJobInput)
Expect(*createdRequest.HyperParameterTuningJobName).To(Equal(controllers.GetGeneratedJobName(tuningJob.ObjectMeta.GetUID(), tuningJob.ObjectMeta.GetName(), MaxHyperParameterTuningJobNameLength)))
```

```go
// from "Spec defines HyperParameterTuningJobName"
req := receivedRequests.Front().Next().Value
Expect(req).To(BeAssignableToTypeOf((*sagemaker.CreateHyperParameterTuningJobInput)(nil)))

createdRequest := req.(*sagemaker.CreateHyperParameterTuningJobInput)
Expect(*createdRequest.HyperParameterTuningJobName).To(Equal("tuning-job-name"))
```

They are identical in everything except the expected job name.